File: 0941-sort-array-by-parity/0941-sort-array-by-parity.java
class Solution {
    public void swap(int arr[],int i,int j){
        int t=arr[i];
        arr[i]=arr[j];
        arr[j]=t;
    }
    public int[] sortArrayByParity(int[] arr) {
        int even=0,odd=0,n=arr.length;
        while(true){
            while(odd<n&&((arr[odd]&1)==0))++odd;
            while(even<n&&(even<odd||((arr[even]&1)==1)))++even;
            if(odd>=n||even>=n)break;
            else swap(arr,even,odd);
        }
        return arr;
    }
}

File: 0941-sort-array-by-parity/0941-sort-array-by-parity.kt
class Solution {
    fun sortArrayByParity(nums: IntArray): IntArray {
        var even=0
        var odd=0
        while(even<nums.size && odd<nums.size){
            if(nums[even] and 1 == 1){
                ++even
                continue
            }
            else if(nums[odd] and 1 == 0){
                ++odd
                continue
            }
            else if(odd>even){
                ++even
                continue
            }
            var t=nums[odd]
            nums[odd]=nums[even]
            nums[even]=t
            ++odd
            ++even
        }
        return nums
    }
}